### ISSUE TITLE
Sending Newsletter from TYPO3 with enforced SSL impossible

### ISSUE THREAD

**Issue by typoworx-de** [Ecodev/newsletter#119]
I noticed some serious problems, while trying to send Newsletter from within a TYPO3 that has enabled lockSSL (backend) and forced ssl for frontend (by native redirect enforced if not using ssl).

It looks like we have two issue's on this:

1. \Ecodev\Newsletter\Utility\Validator::getURL
    The invoked TYPO3-Core-Method GeneralUtility::getURL did not work in my case
    (may be caused by http -> https redirect?!)

2. \Ecodev\Newsletter\Domain\Model\Newsletter::getContentUrl (Model)
    This method *always* returns http-Protocol!

I've applied these two patches resolving the issue. The curl-param disabling ssl-checks was mandatory for our development-system. But I think it won't hurt there anyway.

**Patch for \Ecodev\Newsletter\Utility\Validator::getURL**
```
    protected function getURL($url)
    {
        //return \TYPO3\CMS\Core\Utility\GeneralUtility::getURL($url);

        $ch = curl_init();
        
        $response = null;
        if(isset($ch))
        {
            curl_setopt($ch, CURLOPT_URL, $url);
            curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
            curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, 0);
            curl_setopt($ch, CURLOPT_FAILONERROR, 1);
            //curl_setopt($ch, CURLOPT_FOLLOWLOCATION, 1);
            curl_setopt($ch, CURLOPT_RETURNTRANSFER,1);
            curl_setopt($ch, CURLOPT_TIMEOUT, 8);
            $response = curl_exec($ch);

            if (curl_error($ch))
            {
                $response = null;
            }

            curl_close($ch);
        }
        return $response;
    }
```

**Patch for \Ecodev\Newsletter\Domain\Model\Newsletter::getContentUrl (Model)**

```
    /**
     * Returns the URL of the content of this newsletter
     * @return string
     */
    public function getContentUrl($language = null)
    {
        $append_url = Tools::confParam('append_url');
        $domain = $this->getDomain();

        if (!is_null($language)) {
            $language = '&L=' . $language;
        }

        $isSSL = (!empty($_SERVER['HTTPS']) && $_SERVER['HTTPS'] !== 'off' || $_SERVER['SERVER_PORT'] == 443);
        return sprintf('http%s://%s/index.php?id=%d%s', $isSSL ? 's' : '', $domain, $this->getPid(), $language . $append_url);
    }
```

**Comment by typoworx-de** [comment]
Note:
I did not find a suiteable method in Core to check for HTTPS/Protocol. In case CloudFlare is used, there is some additional code needed in order to detect SSL-Usage!

If there is a fine solution to make GeneralUtility::getURL make follow redirects and work with SSL - I would be glad for any improvements of my patch provided.

Please keep me up-to-date if this patch will be used for future versions of EXT:newsletter.

**Comment by PowerKiKi** [comment]
Upgrade Newsletter, your patch is based on an old version. And remove your
patch. And set the scheme (and domain) in `fetch_path` configuration in
Extension Manager. This should solve your issue... at least the scheme
detection part... not sure what's going on with curl...

**Comment by PowerKiKi** [comment]
Like I said, proper configuration of Newsletter should avoid the wrong scheme issue. As for redirect, I tested on TYPO3 7.6 and 8.1.

On TYPO3 7.6 (or earlier), [cURL always follow redirect](https://github.com/TYPO3/TYPO3.CMS/blob/7.6.9/typo3/sysext/core/Classes/Utility/GeneralUtility.php#L2463), but you have to be sure that cURL is actually used via:

```php
// TYPO3 7.6 or earlier
$GLOBALS['TYPO3_CONF_VARS']['SYS']['curlUse'] = 1; // Use cURL to follow redirects automatically
```

On TYPO3 8.1 [Guzzle was introduced](https://docs.typo3.org/typo3cms/extensions/core/latest/Changelog/8.1/Feature-70056-GuzzleForHttpRequests.html) to replace everything. So you should tell Guzzle to follow redirect. And optionnally ignore invalid SSL certificates, but this not recommended for obvious security concerns:
```php
// TYPO3 8.1
$GLOBALS['TYPO3_CONF_VARS']['HTTP']['allow_redirects'] = true; // Allow redirect via Guzzle
$GLOBALS['TYPO3_CONF_VARS']['HTTP']['verify'] = false; // Disable SSL verification. DOT NOT USE IN PRODUCTION
```

So bottom line is there is nothing to fix in Newsletter AFAIK. Still, in latest version (soon to be released), you will get an error message if content cannot be retrieved. This hopefully should help you better understand what is going on and what should be adjusted.

I will close this for now, but feel free to reopen, if you think something is wrong with Newsletter.

**Comment by typoworx-de** [comment]
@PowerKiKi 
Thanks for your reply. I cannot upgrade Newsletter-Ext as newer versions do not seem to be compatible with the existing TYPO3 6.2 Installation and an Upgrade is currently not possible and wanted by the customer.

With the patch given I got it working on our Development-System (it even works there with an invalid SSL-Cert used on DEV) with the patch given.

But on Production it still has the same issue - same TYPO3-Version and same (patched) Plugin.

Thank you for the note on the fetch_url Parameter in Extension-Manager. If it is recommendet to set it with Protocol-Scheme, then you should update the Text-Note there as it does hint NOT to use the HTTP-Scheme in that configuration-field.
> Domain or domain + path (without "http") from wich to fetch content and encode links with. Leave
> blank to use domain-records from the page tree.

**Comment by typoworx-de** [comment]
Update on using fetch_url with HTTP-Scheme...
It does not work by using it like this: https://myproject.de/

Error still is present. I've also checked your other Configuration-Settings which already have been applied before.

Any other suggestions?
A Module for checking such configuration issues would be awesome... by the side.

**Comment by PowerKiKi** [comment]
This extension support 6.2 as evidenced by [Travis builds](https://travis-ci.org/Ecodev/newsletter/jobs/152995845). I think we should first focus on upgrading the extension. Then revisit the original problem if it still occurs.


The edit suggestion you made about the explanation of `fetch _ path` [was made almost one year ago](https://github.com/Ecodev/newsletter/commit/b8e4e082d909b83eb20f26ce159df955bd3c1957#diff-8ee17d276215f3affcd87ac9221ab664L10). That means the version you use is older than [2.6.0](https://github.com/Ecodev/newsletter/releases/tag/2.6.0). That means that HTTPS is **not** supported in your version.

I won't be able to help you anymore until you upgrade to latest newsletter.

**Comment by totocrenn** [comment]
Hi guys,
I aslo have some trouble to use the newsletter extension with https.
When I go to the newsletter tab and I choose a page, there is some seconds of loading and then get this error in the Newsletter validity --> Errors : 
`Could not fetch "https://www._my-domain_.bzh/index.php?id=45" Error: 28 Message: cURL error 28: Operation timed out after 0 milliseconds with 0 out of 0 bytes received (see http://curl.haxx.se/libcurl/c/libcurl-errors.html)`

When I apply the patch I got an other error : 
`Content too short. The content must be at least 200 chars long to be considered valid.`
My page is far from being under 200 char long...

Any suggestions would be most welcome.

_Newsletter version : 3.2.0
Typo3 version : 8.7.7_

**Comment by PowerKiKi** [comment]
@totocrenn, please open a new issue, while SSL is also involved it seems symptoms are different. And OP is now quite old.
